google/cloud/bigquery_storage_v1/writer.py

```diff
@@ -364,6 +364,291 @@ def _on_rpc_done(self, future):
         thread.start()
 
 
+class Connection(object):
+    def __init__(
+        self,
+        client: big_query_write.BigQueryWriteClient,
+        stream: AppendRowsStream,
+        initial_request_template: gapic_types.AppendRowsRequest,
+        metadata: Sequence[Tuple[str, str]] = (),
+    ) -> None:
+        """A connection abstraction that includes a gRPC connection, a consumer,
+        and a queue.
+
+        Args:
+            client:
+                Client responsible for making requests.
+            stream:
+                The stream that created the connection.
+            initial_request_template:
+                Data to include in the first request sent to the stream. This
+                must contain
+                :attr:`google.cloud.bigquery_storage_v1.types.AppendRowsRequest.write_stream`
+                and
+                :attr:`google.cloud.bigquery_storage_v1.types.AppendRowsRequest.ProtoData.writer_schema`.
+            metadata:
+                Extra headers to include when sending the streaming request.
+        """
+        self._client = client
+        self._stream = stream
+        self._initial_request_template = initial_request_template
+        self._metadata = metadata
+        self._thread_lock = threading.RLock()
+
+        self._rpc = None
+        self._consumer = None
+        self._stream_name = None
+        self._queue = queue.Queue()
+
+        # statuses
+        self._closed = False
+
+    @property
+    def is_active(self) -> bool:
```

Comment:
should this be done with the lock to avoid racy accesses?

Reply:
Thanks, this is a great question. It seems that there was no lock in this method in the original AppendRowsStream class for mainly two reasons:
1. we might need to acquire the lock outside `is_active`, because we may still lose the thread between calling `is_active` and locking the thread inside of it
2. the only places calling `is_active` are `_open()` and `_shutdown()`, which already are using thread lock.

Maybe I can add in the docstring that this needs to be called inside a lock, so at least users and/or future devs are aware, but by no mean is this a guarantee.